libfreerdp/core/info.c

```diff
@@ -724,32 +899,87 @@ BOOL rdp_recv_logon_info_v2(rdpRdp* rdp, wStream* s, logon_info *info)
 	Stream_Read_UINT32(s, cbUserName); /* cbUserName (4 bytes) */
 	Stream_Seek(s, 558); /* pad (558 bytes) */
 
-	if (Stream_GetRemainingLength(s) < (cbDomain + cbUserName))
-		return FALSE;
 
-	if (cbDomain)
+	/* cbDomain is the size in bytes of the Unicode character data in the Domain field.
+	 * The size of the mandatory null terminator is include in this value.
+	 * Note: Since MS-RDPBCGR 2.2.10.1.1.2 does not mention any size limits we assume
+	 *       that the maximum value is 52 bytes, according to the fixed size of the
+	 *       Domain field in the Logon Info Version 1 (TS_LOGON_INFO) structure.
+	 */
+
+	if ((cbDomain % 2) || cbDomain < 2 || cbDomain > 52)
 	{
-		ConvertFromUnicode(CP_UTF8, 0, (WCHAR*) Stream_Pointer(s), cbDomain, &info->domain, 0, NULL, FALSE);
-		if (!info->domain)
-			return FALSE;
-		Stream_Seek(s, cbDomain); /* domain */
+		WLog_ERR(TAG, "protocol error: invalid cbDomain value: %lu", cbDomain);
+		goto fail;
 	}
 
-	if (cbUserName)
+	if (Stream_GetRemainingLength(s) < (size_t) cbDomain)
 	{
-		ConvertFromUnicode(CP_UTF8, 0, (WCHAR*) Stream_Pointer(s), cbUserName, &info->username, 0, NULL, FALSE);
-		if (!info->username)
-		{
-			free(info->domain);
-			info->domain = NULL;
-			return FALSE;
-		}
-		Stream_Seek(s, cbUserName); /* userName */
+		WLog_ERR(TAG, "insufficient remaining stream length");
+		goto fail;
+	}
+
+	wstr = (WCHAR*) Stream_Pointer(s);
+	if (wstr[cbDomain / 2 - 1])
+	{
+		WLog_ERR(TAG, "protocol error: Domain field must be null terminated");
+		goto fail;
 	}
 
-	WLog_DBG(TAG, "LogonInfoV2: SessionId: 0x%04X", info->sessionId);
+	if (ConvertFromUnicode(CP_UTF8, 0, wstr, -1, &info->domain, 0, NULL, FALSE) < 1)
+	{
+		WLog_ERR(TAG, "failed to convert the Domain string");
+		goto fail;
+	}
+
+	Stream_Seek(s, cbDomain); /* domain */
+
+
+	/* cbUserName is the size in bytes of the Unicode character data in the UserName field.
+	 * The size of the mandatory null terminator is include in this value.
+	 * Note: Since MS-RDPBCGR 2.2.10.1.1.2 does not mention any size limits we assume
+	 *       that the maximum value is 512 bytes, according to the fixed size of the
+	 *       Username field in the Logon Info Version 1 (TS_LOGON_INFO) structure.
+	 */
+
+	if ((cbUserName % 2) || cbUserName < 2 || cbUserName > 512)
+	{
```

Comment:
@hardening I cannot reproduce this. And this would be a protocol violation since [MS-RDPBCGR 2.2.10.1.1.2 Logon Info Version 2](https://msdn.microsoft.com/en-us/library/cc240638.aspx) says:
>  cbDomain (4 bytes): A 32-bit, unsigned integer. The size in bytes of the Domain field **(including the mandatory null terminator).**

So IMO the minimum value is 2.  Do you remember how you reproduced that ?